webui/react/src/stores/projects.tsx

```diff
@@ -1,125 +1,120 @@
 import { Map } from 'immutable';
-import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
+import React, {
+  createContext,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
 
 import { getWorkspaceProjects } from 'services/api';
 import { Project } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+import { observable, WritableObservable } from 'utils/observable';
+
+class ProjectService {
+  private _projects: WritableObservable<Map<number, Project>> = observable(Map());
+  private _projectsByIndex: WritableObservable<Map<string, Project[]>> = observable(Map());
+
+  fetchWorkspaceProjects = async (
+    workspaceId: number,
+    canceler: AbortController,
+  ): Promise<void> => {
+    if (this._projectsByIndex.get().get(`byworkspace-${workspaceId}`)) return;
```

Comment:
You should allow callers to specify this behavior since sometimes you _do_ want to refetch. Also you should check this again inside the callback to avoid double updates from racing.

Reply:
I added a `forceFetch` option in case the user want to refetch.
By inside the callback, do you mean right before `this._projectsByIndex.update`? Since we are using map, concurrency should be fine?

Reply:
Yes, before the update. This function is split across two tasks on the event loop. It's possible for multiple invocations of `fetchWorkspaceProjects` to get past this condition if they're scheduled back-to-back on the event loop. You'll then see two API calls made.

It's possible to add a lock to prevent the two separate API calls but we're not using those in the codebase yet. The simpler thing you can do is make sure only one of those API calls actually updates the observable to prevent unecessary re-renders.